Common/src/geometry/CPhysicalGeometry.cpp

```diff
@@ -4346,431 +4346,305 @@ void CPhysicalGeometry::PrepareAdjacency(CConfig *config) {
+        if (checkTetra(Point_1_Surface, Point_2_Surface, Point_3_Surface, Point_Domain)) {
+          bound[iMarker][iElem_Surface]->Change_Orientation();
+          tria_flip++;
+        }
+      }
 
-          for (iDim = 0; iDim < nDim; iDim++) {
-            a[iDim] = 0.5*(Coord_3[iDim]-Coord_2[iDim]);
-            b[iDim] = 0.5*(Coord_4[iDim]-Coord_2[iDim]);
-            c[iDim] = Coord_5[iDim]-Coord_2[iDim];
-          }
-          n[0] = a[1]*b[2]-b[1]*a[2];
-          n[1] = -(a[0]*b[2]-b[0]*a[2]);
-          n[2] = a[0]*b[1]-b[0]*a[1];
-          test_2 = n[0]*c[0]+n[1]*c[1]+n[2]*c[2];
+      if (bound[iMarker][iElem_Surface]->GetVTK_Type() == QUADRILATERAL) {
 
-          for (iDim = 0; iDim < nDim; iDim++) {
-            a[iDim] = 0.5*(Coord_4[iDim]-Coord_3[iDim]);
-            b[iDim] = 0.5*(Coord_1[iDim]-Coord_3[iDim]);
-            c[iDim] = Coord_5[iDim]-Coord_3[iDim];
-          }
-          n[0] = a[1]*b[2]-b[1]*a[2];
-          n[1] = -(a[0]*b[2]-b[0]*a[2]);
-          n[2] = a[0]*b[1]-b[0]*a[1];
-          test_3 = n[0]*c[0]+n[1]*c[1]+n[2]*c[2];
+        auto Point_1_Surface = bound[iMarker][iElem_Surface]->GetNode(0);
+        auto Point_2_Surface = bound[iMarker][iElem_Surface]->GetNode(1);
+        auto Point_3_Surface = bound[iMarker][iElem_Surface]->GetNode(2);
+        auto Point_4_Surface = bound[iMarker][iElem_Surface]->GetNode(3);
 
-          for (iDim = 0; iDim < nDim; iDim++) {
-            a[iDim] = 0.5*(Coord_1[iDim]-Coord_4[iDim]);
-            b[iDim] = 0.5*(Coord_3[iDim]-Coord_4[iDim]);
-            c[iDim] = Coord_5[iDim]-Coord_4[iDim];
-          }
-          n[0] = a[1]*b[2]-b[1]*a[2];
-          n[1] = -(a[0]*b[2]-b[0]*a[2]);
-          n[2] = a[0]*b[1]-b[0]*a[1];
-          test_4 = n[0]*c[0]+n[1]*c[1]+n[2]*c[2];
-
-          if ((test_1 < 0.0) && (test_2 < 0.0) && (test_3 < 0.0) && (test_4 < 0.0)) {
-            bound[iMarker][iElem_Surface]->Change_Orientation();
-            quad_flip++;
-          }
+        /*--- Divide quadrilateral/pyramid into triangles/tetrahedrons. ---*/
+        int test_1 = checkTetra(Point_1_Surface, Point_2_Surface, Point_3_Surface, Point_Domain);
+        int test_2 = checkTetra(Point_2_Surface, Point_3_Surface, Point_4_Surface, Point_Domain);
+        int test_3 = checkTetra(Point_3_Surface, Point_4_Surface, Point_1_Surface, Point_Domain);
+        int test_4 = checkTetra(Point_4_Surface, Point_1_Surface, Point_2_Surface, Point_Domain);
 
+        if (test_1+test_2+test_3+test_4 >= 3) {
+          /*--- If 3 or 4 tests fail there is > 75% chance flipping is the right choice. ---*/
+          bound[iMarker][iElem_Surface]->Change_Orientation();
+          quad_flip++;
+        }
+        else if (test_1+test_2+test_3+test_4 == 2) {
+          /*--- If 50/50 we cannot be sure of what to do -> report to user.
+           * If only one test fails it is probably (75%) due to skewness or warping. ---*/
+          quad_error++;
```

Comment:
It is probably not the best way, but a bullet proof check probably needs to go over many combinations of nodes, as for making it an hard error, I don't like the idea very much because the skewness might by due to an hard to mesh geometry, and I don't wan't people to complain I broke their cases.